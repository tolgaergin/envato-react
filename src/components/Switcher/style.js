import styled from 'styled-components';

const theme = {
  rowWidth: '52',
  rowHeight: '32',
  rowBgColor: 'rgba(0, 0, 0, 0.2)',
  rowRadius: '20',

  thumbWidth: '26',
  thumbHeight: '26',
  thumbLeftSpace: '3',
  thumbRightSpace: '3',
  thumbBgColor: '#fff',
  thumbBoxShadow: '0 2px 4px rgba(0, 0, 0, 0.4)',
  pressedThumbBoxShadow: '0 2px 8px rgba(0, 0, 0, 0.3), 0 0 0 20px rgba(128, 128, 128, 0.1)',

  activeRowBgColor: '#37C34A',
  activeThumbBgColor: '#fff',
  activeThumbBoxShadow: '0 3px 2px rgba(0, 0, 0, 0.05)',
  activePressedThumbBoxShadow: '0 2px 8px rgba(0, 0, 0, 0.3), 0 0 0 20px rgba(55, 195, 74, 0.2)',
};

export const StyledSwitch = styled.div `
  display: inline-block;
  position: relative;
  width: ${theme.rowWidth}px;
  height: ${theme.rowHeight}px;
  margin: 0;
  border-radius: ${theme.rowRadius}px;
  background-color: ${theme.rowBgColor};
  transition: background-color 280ms ease-out;
  cursor: pointer;

  &::before {
    content: '';
    position: absolute;
    top: ${(theme.rowHeight - theme.thumbHeight) / 2}px;
    left: ${theme.thumbLeftSpace}px;
    width: ${theme.thumbWidth}px;
    height: ${theme.thumbHeight}px;
    background: ${theme.thumbBgColor};
    box-shadow: ${theme.thumbBoxShadow};
    border-radius: 50%;
    transition: all 280ms ease-out;
  }

  &:active {
    &::before {
      box-shadow: ${theme.pressedThumbBoxShadow};
    }
  }
`;

export const StyledInput = styled.input `
  display: none;

  &:checked + div {
    background: ${theme.activeRowBgColor};

    &::before {
      left: ${theme.rowWidth - theme.thumbWidth - theme.thumbRightSpace}px;
      background: ${theme.activeThumbBgColor};
      box-shadow: ${theme.activeThumbBoxShadow};
    }

    &:active {
      &::before {
        box-shadow: ${theme.activePressedThumbBoxShadow};
      }
    }
  }
`;
