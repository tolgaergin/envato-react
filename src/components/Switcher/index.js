import React from 'react';
import { StyledInput, StyledSwitch } from './style';

const Switcher = ({ name, checked, onChange }) => (
  <label>
    <StyledInput
      type="checkbox"
      name={name}
      checked={checked}
      onChange={onChange} />
    <StyledSwitch />
  </label>
);

Switcher.propTypes = {
  checked: React.PropTypes.bool.isRequired,
  name: React.PropTypes.string,
  onChange: React.PropTypes.func.isRequired,
};

export default Switcher;
