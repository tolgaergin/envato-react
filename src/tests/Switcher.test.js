import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';

import Switcher from '../components/Switcher/index';

describe('Switcher', () => {
  const checked = true;
  const name = 'switcher';
  const onChange = jest.fn();

  it('renders and matches snapshot', () => {
    const component = renderer.create(
      <Switcher checked={checked} name={name} onChange={onChange} />
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('button click calls onAdd', () => {
    // manually trigger the callback
    const component = mount(
      <Switcher checked={checked} name={name} onChange={onChange} />
    );

    const container = component.find('input');
    expect(container.length).toEqual(1);

    component.find('input').simulate('change');
    expect(onChange).toBeCalled();
  });
});
