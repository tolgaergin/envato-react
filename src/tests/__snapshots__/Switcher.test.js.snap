exports[`Switcher renders and matches snapshot 1`] = `
<label>
  <input
    checked={true}
    className="cQZeVv"
    name="switcher"
    onChange={[Function]}
    type="checkbox" />
  <div
    className="iVGvmF" />
</label>
`;
